File: demo-java/app/src/main/java/com/example/placesdemo/PlaceAutocompleteActivity.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.placesdemo;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.example.placesdemo.databinding.PlaceAutocompleteActivityBinding;
import com.google.android.gms.common.api.Status;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.tasks.Task;
import com.google.android.libraries.places.api.Places;
import com.google.android.libraries.places.api.model.AutocompleteSessionToken;
import com.google.android.libraries.places.api.model.LocationBias;
import com.google.android.libraries.places.api.model.LocationRestriction;
import com.google.android.libraries.places.api.model.Place;
import com.google.android.libraries.places.api.model.RectangularBounds;
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest;
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsResponse;
import com.google.android.libraries.places.api.net.PlacesClient;
import com.google.android.libraries.places.widget.Autocomplete;
import com.google.android.libraries.places.widget.AutocompleteActivity;
import com.google.android.libraries.places.widget.AutocompleteSupportFragment;
import com.google.android.libraries.places.widget.listener.PlaceSelectionListener;
import com.google.android.libraries.places.widget.model.AutocompleteActivityMode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.activity.EdgeToEdge;

/**
 * Activity to demonstrate Place Autocomplete (activity widget intent, fragment widget, and
 * {@link PlacesClient#([PlacesClient.findAutocompletePredictions])}).
 */
public class PlaceAutocompleteActivity extends AppCompatActivity {

    private static final int AUTOCOMPLETE_REQUEST_CODE = 23487;
    private PlacesClient placesClient;
    private FieldSelector fieldSelector;
    private PlaceAutocompleteActivityBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        // Enable edge-to-edge display. This must be called before calling super.onCreate().
        EdgeToEdge.enable(this);
        super.onCreate(savedInstanceState);

        setContentView(R.layout.place_autocomplete_activity);

        binding = PlaceAutocompleteActivityBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        // Retrieve a PlacesClient (previously initialized - see MainActivity)
        placesClient = Places.createClient(this);

        // Set up view objects
        binding.autocompleteUseTypesFilterCheckbox.setOnCheckedChangeListener(
                (buttonView, isChecked) -> binding.autocompleteTypesFilterEdittext.setEnabled(isChecked));
        fieldSelector =
                new FieldSelector(
                        binding.useCustomFields,
                        binding.customFieldsList,
                        savedInstanceState);

        setupAutocompleteSupportFragment();

        // Set listeners for Autocomplete activity
        binding.autocompleteActivityButton
                .setOnClickListener(view -> startAutocompleteActivity());

        // Set listeners for programmatic Autocomplete
        binding.fetchAutocompletePredictionsButton.setOnClickListener(view -> findAutocompletePredictions());

        // UI initialization
        setLoading(false);
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle bundle) {
        super.onSaveInstanceState(bundle);
        fieldSelector.onSaveInstanceState(bundle);
    }

    private void setupAutocompleteSupportFragment() {
        final AutocompleteSupportFragment autocompleteSupportFragment =
                (AutocompleteSupportFragment)
                        getSupportFragmentManager().findFragmentById(R.id.autocomplete_support_fragment);
        if (autocompleteSupportFragment != null) {
            autocompleteSupportFragment.setPlaceFields(getPlaceFields());
            autocompleteSupportFragment.setOnPlaceSelectedListener(getPlaceSelectionListener());
        }

        binding.autocompleteSupportFragmentUpdateButton
                .setOnClickListener(
                        view ->
                                autocompleteSupportFragment
                                        .setPlaceFields(getPlaceFields())
                                        .setText(getQuery())
                                        .setHint(getHint())
                                        .setCountries(getCountries())
                                        .setLocationBias(getLocationBias())
                                        .setLocationRestriction(getLocationRestriction())
                                        .setTypesFilter(getTypesFilter())
                                        .setActivityMode(getMode()));
    }

    private PlaceSelectionListener getPlaceSelectionListener() {
        return new PlaceSelectionListener() {
            @Override
            public void onPlaceSelected(@NonNull Place place) {
                binding.response.setText(
                        StringUtil.stringifyAutocompleteWidget(place, isDisplayRawResultsChecked()));
            }

            @Override
            public void onError(@NonNull Status status) {
                binding.response.setText(status.getStatusMessage());
            }
        };
    }

    /**
     * Called when AutocompleteActivity finishes
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent intent) {
        if (requestCode == AUTOCOMPLETE_REQUEST_CODE) {
            if (resultCode == AutocompleteActivity.RESULT_OK) {
                Place place = Autocomplete.getPlaceFromIntent(intent);
                binding.response.setText(
                        StringUtil.stringifyAutocompleteWidget(place, isDisplayRawResultsChecked()));
            } else if (resultCode == AutocompleteActivity.RESULT_ERROR) {
                Status status = Autocomplete.getStatusFromIntent(intent);
                binding.response.setText(status.getStatusMessage());
            }  // The user canceled the operation.

        }

        // Required because this class extends AppCompatActivity which extends FragmentActivity
        // which implements this method to pass onActivityResult calls to child fragments
        // (eg AutocompleteFragment).
        super.onActivityResult(requestCode, resultCode, intent);
    }

    private void startAutocompleteActivity() {
        Intent autocompleteIntent =
                new Autocomplete.IntentBuilder(getMode(), getPlaceFields())
                        .setInitialQuery(getQuery())
                        .setHint(getHint())
                        .setCountries(getCountries())
                        .setLocationBias(getLocationBias())
                        .setLocationRestriction(getLocationRestriction())
                        .setTypesFilter(getTypesFilter())
                        .build(this);
        startActivityForResult(autocompleteIntent, AUTOCOMPLETE_REQUEST_CODE);
    }

    private void findAutocompletePredictions() {
        setLoading(true);

        FindAutocompletePredictionsRequest.Builder requestBuilder =
                FindAutocompletePredictionsRequest.builder()
                        .setQuery(getQuery())
                        .setCountries(getCountries())
                        .setOrigin((getOrigin()))
                        .setLocationBias(getLocationBias())
                        .setLocationRestriction(getLocationRestriction())
                        .setTypesFilter(getTypesFilter());

        if (isUseSessionTokenChecked()) {
            requestBuilder.setSessionToken(AutocompleteSessionToken.newInstance());
        }

        Task<FindAutocompletePredictionsResponse> task =
                placesClient.findAutocompletePredictions(requestBuilder.build());

        task.addOnSuccessListener(
                (response) ->
                        binding.response.setText(StringUtil.stringify(response, isDisplayRawResultsChecked())));

        task.addOnFailureListener(
                (exception) -> {
                    exception.printStackTrace();
                    binding.response.setText(exception.getMessage());
                });

        task.addOnCompleteListener(response -> setLoading(false));
    }

    //////////////////////////
    // Helper methods below //
    //////////////////////////

    private List<Place.Field> getPlaceFields() {
        if (((CheckBox) findViewById(R.id.use_custom_fields)).isChecked()) {
            return fieldSelector.getSelectedFields();
        } else {
            return fieldSelector.getAllFields();
        }
    }

    @Nullable
    private String getQuery() {
        return getTextViewValue(R.id.autocomplete_query);
    }

    @Nullable
    private String getHint() {
        return getTextViewValue(R.id.autocomplete_hint);
    }

    private List<String> getCountries() {
        String countryString = getTextViewValue(R.id.autocomplete_country);
        if (TextUtils.isEmpty(countryString)) {
            return new ArrayList<>();
        }

        return StringUtil.countriesStringToArrayList(countryString);
    }

    @Nullable
    private String getTextViewValue(@IdRes int textViewResId) {
        String value = ((TextView) findViewById(textViewResId)).getText().toString();
        return TextUtils.isEmpty(value) ? null : value;
    }

    @Nullable
    private LocationBias getLocationBias() {
        return getBounds(
                R.id.autocomplete_location_bias_south_west, R.id.autocomplete_location_bias_north_east);
    }

    @Nullable
    private LocationRestriction getLocationRestriction() {
        return getBounds(
                R.id.autocomplete_location_restriction_south_west,
                R.id.autocomplete_location_restriction_north_east);
    }

    @Nullable
    private RectangularBounds getBounds(int resIdSouthWest, int resIdNorthEast) {
        String southWest = ((TextView) findViewById(resIdSouthWest)).getText().toString();
        String northEast = ((TextView) findViewById(resIdNorthEast)).getText().toString();
        if (TextUtils.isEmpty(southWest) && TextUtils.isEmpty(northEast)) {
            return null;
        }

        LatLngBounds bounds = StringUtil.convertToLatLngBounds(southWest, northEast);
        if (bounds == null) {
            showErrorAlert(R.string.error_alert_message_invalid_bounds);
            return null;
        }

        return RectangularBounds.newInstance(bounds);
    }

    @Nullable
    private LatLng getOrigin() {
        String originStr =
                ((TextView) findViewById(R.id.autocomplete_location_origin)).getText().toString();
        if (TextUtils.isEmpty(originStr)) {
            return null;
        }

        LatLng origin = StringUtil.convertToLatLng(originStr);
        if (origin == null) {
            showErrorAlert(R.string.error_alert_message_invalid_origin);
            return null;
        }

        return origin;
    }

    private List<String> getTypesFilter() {
        EditText typesFilterEditText = findViewById(R.id.autocomplete_types_filter_edittext);
        return typesFilterEditText.isEnabled()
                ? Arrays.asList(typesFilterEditText.getText().toString().split("[\\s,]+"))
                : new ArrayList<>();
    }


    private AutocompleteActivityMode getMode() {
        boolean isOverlayMode =
                ((CheckBox) findViewById(R.id.autocomplete_activity_overlay_mode)).isChecked();
        return isOverlayMode ? AutocompleteActivityMode.OVERLAY : AutocompleteActivityMode.FULLSCREEN;
    }

    private boolean isDisplayRawResultsChecked() {
        return ((CheckBox) findViewById(R.id.display_raw_results)).isChecked();
    }

    private boolean isUseSessionTokenChecked() {
        return ((CheckBox) findViewById(R.id.autocomplete_use_session_token)).isChecked();
    }

    private void setLoading(boolean loading) {
        findViewById(R.id.loading).setVisibility(loading ? View.VISIBLE : View.INVISIBLE);
    }

    private void showErrorAlert(@StringRes int messageResId) {
        new AlertDialog.Builder(this)
                .setTitle(R.string.error_alert_title)
                .setMessage(messageResId)
                .show();
    }
}

File: demo-kotlin/app/src/main/java/com/example/placesdemo/PlaceAutocompleteActivity.kt
/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.placesdemo

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.CheckBox
import android.widget.TextView
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.IdRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.example.placesdemo.databinding.PlaceAutocompleteActivityBinding
import com.google.android.gms.common.api.Status
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.LocationBias
import com.google.android.libraries.places.api.model.LocationRestriction
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.model.RectangularBounds
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsResponse
import com.google.android.libraries.places.api.net.PlacesClient
import com.google.android.libraries.places.widget.PlaceAutocompleteActivity
import com.google.android.libraries.places.widget.AutocompleteSupportFragment
import com.google.android.libraries.places.widget.PlaceAutocomplete
import com.google.android.libraries.places.widget.listener.PlaceSelectionListener
import com.google.android.libraries.places.widget.model.AutocompleteActivityMode

/**
 * Activity to demonstrate Place Autocomplete (activity widget intent, fragment widget, and
 * [PlacesClient.findAutocompletePredictions]).
 */
class PlaceAutocompleteActivity : BaseActivity() {

    private lateinit var placesClient: PlacesClient
    private lateinit var fieldSelector: FieldSelector

    private lateinit var binding: PlaceAutocompleteActivityBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = PlaceAutocompleteActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setSupportActionBar(binding.topBar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        binding.topBar.setNavigationOnClickListener {
            onBackPressedDispatcher.onBackPressed()
        }

        // Retrieve a PlacesClient (previously initialized - see MainActivity)
        placesClient = Places.createClient(this)

        // Set up view objects

        val useTypesFilterCheckBox =
            findViewById<CheckBox>(R.id.autocomplete_use_types_filter_checkbox)
        useTypesFilterCheckBox.setOnCheckedChangeListener { _, isChecked: Boolean ->
            binding.autocompleteTypesFilterEdittext.isEnabled = isChecked
        }
        fieldSelector = FieldSelector(
            findViewById(R.id.use_custom_fields),
            findViewById(R.id.custom_fields_list),
            savedInstanceState
        )
        setupAutocompleteSupportFragment()

        // Set listeners for Autocomplete activity
        binding.autocompleteActivityButton
            .setOnClickListener { startAutocompleteActivity() }

        // Set listeners for programmatic Autocomplete
        binding.fetchAutocompletePredictionsButton
            .setOnClickListener { findAutocompletePredictions() }

        // UI initialization
        setLoading(false)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        fieldSelector.onSaveInstanceState(outState)
    }

    private fun setupAutocompleteSupportFragment() {
        val autocompleteSupportFragment =
            supportFragmentManager.findFragmentById(R.id.autocomplete_support_fragment) as AutocompleteSupportFragment?
        autocompleteSupportFragment!!.setPlaceFields(placeFields)
        autocompleteSupportFragment.setOnPlaceSelectedListener(placeSelectionListener)
        findViewById<View>(R.id.autocomplete_support_fragment_update_button)
            .setOnClickListener {
                autocompleteSupportFragment
                    .setPlaceFields(placeFields)
                    .setText(query)
                    .setHint(hint)
                    .setCountries(countries)
                    .setLocationBias(locationBias)
                    .setLocationRestriction(locationRestriction)
                    .setTypesFilter(getTypesFilter())
                    .setActivityMode(mode)
            }
    }

    private val placeSelectionListener: PlaceSelectionListener
        get() = object : PlaceSelectionListener {
            override fun onPlaceSelected(place: Place) {
                binding.response.text =
                    StringUtil.stringifyAutocompleteWidget(place, isDisplayRawResultsChecked)
            }

            override fun onError(status: Status) {
                binding.response.text = status.statusMessage
            }
        }

    /**
     * Launches Autocomplete activity and handles result
     */
    private var autocompleteLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
        result ->
        when (result.resultCode) {
            RESULT_OK -> {
                val data: Intent? = result.data
                if (data != null) {
                    val place = PlaceAutocomplete.getPredictionFromIntent(data)
                    binding.response.text =
                        StringUtil.stringifyAutocompletePrediction(place, isDisplayRawResultsChecked)
                }
            }
            PlaceAutocompleteActivity.RESULT_ERROR -> {
                val status = PlaceAutocomplete.getResultStatusFromIntent(intent)
                binding.response.text = status?.statusMessage
            }
            RESULT_CANCELED -> {
                // The user canceled the operation.
            }
        }
    }

    private fun startAutocompleteActivity() {
        val autocompleteIntent = PlaceAutocomplete.IntentBuilder()
            .setInitialQuery(query)
            .setCountries(countries)
            .setLocationBias(locationBias)
            .setLocationRestriction(locationRestriction)
            .setTypesFilter(getTypesFilter())
            .build(this)
        autocompleteLauncher.launch(autocompleteIntent)
    }

    private fun findAutocompletePredictions() {
        setLoading(true)
        val requestBuilder = FindAutocompletePredictionsRequest.builder()
            .setQuery(query)
            .setCountries(countries)
            .setOrigin(origin)
            .setLocationBias(locationBias)
            .setLocationRestriction(locationRestriction)
            .setTypesFilter(getTypesFilter())
        if (isUseSessionTokenChecked) {
            requestBuilder.sessionToken = AutocompleteSessionToken.newInstance()
        }
        val task = placesClient.findAutocompletePredictions(requestBuilder.build())
        task.addOnSuccessListener { response: FindAutocompletePredictionsResponse? ->
            response?.let {
                binding.response.text = StringUtil.stringify(it, isDisplayRawResultsChecked)
            }
        }
        task.addOnFailureListener { exception: Exception ->
            exception.printStackTrace()
            binding.response.text = exception.message
        }
        task.addOnCompleteListener {
            setLoading(
                false
            )
        }
    }

    //////////////////////////
    // Helper methods below //
    //////////////////////////
    private val placeFields: List<Place.Field>
        get() = if ((findViewById<View>(R.id.use_custom_fields) as CheckBox).isChecked) {
            fieldSelector.selectedFields
        } else {
            fieldSelector.allFields
        }

    private val query: String?
        get() = getTextViewValue(R.id.autocomplete_query)

    private val hint: String?
        get() = getTextViewValue(R.id.autocomplete_hint)

    private val countries: List<String>
        get() {
            val countryString = getTextViewValue(R.id.autocomplete_country) ?: return emptyList()
            return StringUtil.countriesStringToArrayList(countryString)
        }

    private fun getTextViewValue(@IdRes textViewResId: Int): String? {
        val value = (findViewById<View>(textViewResId) as TextView).text.toString()
        return if (TextUtils.isEmpty(value)) null else value
    }

    private val locationBias: LocationBias?
        get() = getBounds(
            R.id.autocomplete_location_bias_south_west, R.id.autocomplete_location_bias_north_east
        )

    private val locationRestriction: LocationRestriction?
        get() = getBounds(
            R.id.autocomplete_location_restriction_south_west,
            R.id.autocomplete_location_restriction_north_east
        )

    private fun getBounds(resIdSouthWest: Int, resIdNorthEast: Int): RectangularBounds? {
        val southWest = findViewById<TextView>(resIdSouthWest).text.toString()
        val northEast = findViewById<TextView>(resIdNorthEast).text.toString()
        if (TextUtils.isEmpty(southWest) && TextUtils.isEmpty(northEast)) {
            return null
        }
        val bounds = StringUtil.convertToLatLngBounds(southWest, northEast)
        if (bounds == null) {
            showErrorAlert(R.string.error_alert_message_invalid_bounds)
            return null
        }
        return RectangularBounds.newInstance(bounds)
    }

    private val origin: LatLng?
        get() {
            val originStr =
                findViewById<TextView>(R.id.autocomplete_location_origin).text.toString()
            if (TextUtils.isEmpty(originStr)) {
                return null
            }
            val origin = StringUtil.convertToLatLng(originStr)
            if (origin == null) {
                showErrorAlert(R.string.error_alert_message_invalid_origin)
                return null
            }
            return origin
        }

    private fun getTypesFilter(): List<String> {
        return if (binding.autocompleteTypesFilterEdittext.isEnabled)
            binding.autocompleteTypesFilterEdittext.text.toString().split("[\\s,]+".toRegex())
        else emptyList()
    }

    // This Enum is deprecated, but there is no replacement. See https://developers.google.com/maps/documentation/places/android-sdk/reference/com/google/android/libraries/places/widget/model/AutocompleteActivityMode
    private val mode: AutocompleteActivityMode
        get() {
            val isOverlayMode =
                binding.autocompleteActivityOverlayMode.isChecked
            return if (isOverlayMode) AutocompleteActivityMode.OVERLAY else AutocompleteActivityMode.FULLSCREEN
        }

    private val isDisplayRawResultsChecked: Boolean
        get() = binding.displayRawResults.isChecked

    private val isUseSessionTokenChecked: Boolean
        get() = binding.autocompleteUseSessionToken.isChecked

    private fun setLoading(loading: Boolean) {
        findViewById<View>(R.id.loading).visibility = if (loading) View.VISIBLE else View.INVISIBLE
    }

    private fun showErrorAlert(@StringRes messageResId: Int) {
        AlertDialog.Builder(this)
            .setTitle(R.string.error_alert_title)
            .setMessage(messageResId)
            .show()
    }
}